### ISSUE TITLE
Non-deterministic ordering of printed component attributes

### ISSUE THREAD

**Issue by keckler** [terrapower/armi#438]
I am writing ARMI cases programmatically, and then as each case is created, using the `writeInputs` method to print the blueprints file. I happened to notice that each time the blueprints file is written, the ordering of written component attributes (i.e. material, temperature, dimensions) changes, even when nothing has changed between the cases. Even if I just write the same case twice by executing my scripts twice in a row, the ordering may change. 

It would be nice for the ordering to stay the same so that the written input files can be easily `diff`-ed against each other. 

A quick look into ARMI makes me think that this is actually an issue with `yamlize`, but confirmation of that would be helpful. This seems odd, considering Python [now enforces dictionaries to be insertion-ordered](https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6). But if it is something within ARMI, my feeling is that it would be beneficial to update, if possible.

**Comment by youngmit** [comment]
This is something that we used to do, but stopped when we declared minimum sorry for python 3.6, which switched to deterministic dictionary ordering. However, sets may still be ordered stochastically.... 🤔 Might be that something is getting put into a set. It's that maybe something you are doing in your blueprint generation?

**Comment by keckler** [comment]
I do not use sets in my scripts that generate the blueprints. 

But I did a quick scan through `yamlize` and found that it uses sets in a number of places:
- https://github.com/SimplyKnownAsG/yamlize/blob/master/yamlize/maps.py#L12
- https://github.com/SimplyKnownAsG/yamlize/blob/master/yamlize/objects.py#L211
- https://github.com/SimplyKnownAsG/yamlize/blob/master/yamlize/objects.py#L211
- https://github.com/SimplyKnownAsG/yamlize/blob/master/yamlize/yamlizable.py#L19

Just briefly looking at those sets in `yamlize`, it seems likely to me that that is the source of the randomness.

**Comment by john-science** [comment]
Well, looking at @keckler 's references above, this seems like a problem that would be significantly easier to fix in `yamlize`.  We could subclass a million things, and try to fix the API in `armi`, but I am doubtful that would be any fun.

I will give this a whack and try to fix this in `yamlize`.  It might turn out to be quite easy to fix. But fixing the ordering might turn out to have side-effects that break things in the library.  Who knows.

What I can tell you, is fixing the ordering of a set and a dict is much easier in Python 3.7+ than before that.

**Comment by john-science** [comment]
I have opened a ticket for deterministic ordered in `yamlize`: https://github.com/SimplyKnownAsG/yamlize/issues/12

And I will probably do the work myself, I'm just not absolutely certain I see all the places changes would need to be made. Yet.

**Comment by john-science** [comment]
Hey, @keckler, can you send me an example or two of this problem?

You can do it via email, or whatever, so it doesn't have to be cleaned up for public consumption. But I want something to test against.  The `yamlize` project does actually have a lot of tooling for keeping the ordering sound. (Those sets above aren't used in the ordering in the library.)  So there are going to be some finer points in `yamlize` for me to sort out the problem.

Anyway, having working examples to test against would probably help speed up the process.

Thanks!

**Comment by keckler** [comment]
Yeah, I will send you something via email. Sorry about there not being one in the first place.

**Comment by john-science** [comment]
I have a unit test that demonstrates this bug:

https://github.com/john-science/armi/commit/b075b5719cc0ef9d555e863fe3c8cbf1ed396a56

Essentially, the test always fails, but it wouldn't if we had deterministic ordering.

In particular, the attributes of the given block are printed in a different random order every time I run the test.

**Comment by john-science** [comment]
Okay, so this is kind of our problem.

Essentially, Yamlize is really good about deterministically ordering elements before they get written out to the YAML file.

But we go out of our way to edit the items that go into the YAML file on the fly, after Yamlize does all of it's setup and organization. For instance, in all the methods called `addDegreeOfFreedom()` in `suiteBuilder.py`:

https://github.com/terrapower/armi/blob/49ae3fd7b1a2934dbf0f642d9429bb4b1c75ab99/armi/cases/suiteBuilder.py#L274-L277

I _can_ say that I have made a TON of YAML files with our code testing this bug, and they are very largely deterministically ordered.  You have to get 4 levels down into the data heirachy inside them to start seeing non-deterministic ordering.

I am looking to see if there is a low-time-cost way of solving this.

**Comment by john-science** [comment]
Okay, after more trial and error, I can't find a great solution to this problem.

I tried to access the inner workings of our three YAML libraries, to fix the problem at it's source: `yamlize`, 'pyyaml`, and 'ruamel`. But the problem appears to be that changing the YAML mid-stream is the issue.

I thought I could fix this by forcing the output YAML file to be in alphabetic order:

```python
import yaml
f = open('file_path.yaml', 'r')
doc = yaml.load(f, Loader=yaml.FullLoader)
yaml.dump(doc, sort_keys=True)
```

or

```python
from ruamel.yaml import YAML
yaml = YAML()
yaml.dump(custom_yaml_sorter(yamlString), sys.stdout)
```

But both of these features fail pretty hard in ARMI, because of our use of YAML anchors.

I have burnt too much time on this for now, so I'm going to put it down.  I think the real solution is to change the order we do things, so the `Yamlize` object is formed _after_ we make calls to things like `addDegreeOfFreedom()`, not before.

**Comment by john-science** [comment]
I believe this ticket may be complete and is closable.

A long time ago, we changed the number and IDs of all objects in an ARMI reactor: https://github.com/terrapower/armi/pull/1383

This logic only kicks into place once an object is put into a reactor, but then everything should be extremely predictable.

**Comment by keckler** [comment]
I'd like to highlight that this is still an issue. And it cost us significant time recently, since:
1. It is not possible to diff yaml files against each other if they are not deterministically ordered
2. It is difficult to programmatically diff the yaml in memory because it is nested at many levels with lists thrown in there

**Comment by keckler** [comment]
Can we restart the conversation on this? I can set the `PYTHONHASHSEED` in my environment, but this is abnormal and confusing to other people, and it would be great if we could have a unified solution.

**Comment by john-science** [comment]
We should talk about this. Frankly, the YAML part worries me a little, as we don't own either of the YAML libraries we use.

BUT, we own the rest of the code and the only reason I haven't done this yet is priorities and available time. I'd be happy to work on this now.

**Comment by john-science** [comment]
@keckler came up with the idea of solving this problem downstream at the application level. So, instead of setting `PYTHONHASHSEED` in ARMI, the downstream ARMI-based application could set it. (We could even document this as best practice in ARMI.)

Other ideas:

1. Put it in your personal `.bashrc` file. This would be easy and work, but would put the responsibility onto the user.
2. ARMI could make a setting that if left default would do nothing, but could be used to set the `PYTHONHASHSEED`

Do you have any preferences? @keckler ? @opotowsky ?

**Comment by opotowsky** [comment]
Downstream we could do this in two ways: add the env variable to the app init (prob not a best practice, but would gain us the ability to not worry about it being set on different systems), or add it to the app modulefile (for which there will be a dev and prod version).

**Comment by keckler** [comment]
I think setting the `PYTHONHASHSEED` at the application level is a workaround that will work, and sounds like Arrielle is fine with doing this for the application that I care about. So that would solve my problem.

I wouldn't close this ticket based on that, though, since it doesn't solve the issue for everyone. As far as I'm concerned, this is still an issue at the ARMI level and would continue to impact other ARMI users outside of TerraPower.

**Comment by opotowsky** [comment]
> I think setting the `PYTHONHASHSEED` at the application level is a workaround that will work, and sounds like Arrielle is fine with doing this for the application that I care about. So that would solve my problem.

I'm happy to do this unless we do end up putting the solution into ARMI

> I wouldn't close this ticket based on that, though, since it doesn't solve the issue for everyone. As far as I'm concerned, this is still an issue at the ARMI level and would continue to impact other ARMI users outside of TerraPower.

Should we just fix yamlize to handle this for us? Could it get done in time for the next release?
